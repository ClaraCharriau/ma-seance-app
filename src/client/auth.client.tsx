/* eslint-disable  @typescript-eslint/no-explicit-any */
import { User } from '../models/User';
import axios, { AxiosResponse } from 'axios';
import { HOST, handleDeleteResponse, handleError, handleResponse } from './client.utils';

const user_auth_path = HOST + '/auth';
const user_verify_path = HOST + '/verify';
const user_sign_in_path = HOST + '/sign-in';
const user_sign_out_path = HOST + '/sign-out';

export const loginUser = async (email: string, password: string): Promise<User> => {
    return await axios
        .post(user_auth_path, {
            email: email,
            password: password
        })
        .then(response => handleResponse(response))
        .catch(error => handleError(error));
};

export const checkAccountExists = async (email: string): Promise<boolean> => {
    return await axios
        .post(user_verify_path, {
            email: email
        })
        .then(response => {
            console.log('user-verify-client-response POST verify :', response.data);
            if (response.data && typeof response.data.exists === 'boolean') {
                return response.data.exists;
            }
            throw Error('Could not parse api /verify response. ');
        })
        .catch(error => handleError(error));
};

export const signIn = async (pseudo: string, email: string, password: string): Promise<User> => {
    return await axios
        .post(user_sign_in_path, {
            pseudo: pseudo,
            email: email,
            password: password
        })
        .then(response => handleResponse(response))
        .catch(error => handleError(error));
};

export const updateAccount = async (pseudo: string, email: string, password: string): Promise<User> => {
    return await axios
        .patch(user_sign_in_path, {
            pseudo: pseudo,
            email: email,
            password: password
        })
        .then(response => handleResponse(response))
        .catch(error => handleError(error));
};

export const deleteAccount = async (id: number): Promise<AxiosResponse<any, any>> => {
    return await axios
        .delete(user_sign_out_path + '/' + id)
        .then(response => handleDeleteResponse(response))
        .catch(error => handleError(error));
};
