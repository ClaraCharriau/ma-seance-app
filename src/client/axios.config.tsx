import axios, { AxiosResponse } from 'axios';
import { toast } from 'react-toastify';
import config from '../config/config.helper';
import { UserToken } from '../model/UserToken';

/**
 * Set axios global configuration
 */
export const axiosInstance = axios.create({
    baseURL: config.services.prefix,
    headers: {
        'Content-Type': 'application/json'
    }
});

/**
 * Set authentication header if token is found
 */
axiosInstance.interceptors.request.use(config => {
    const maSeanceIdToken = localStorage.getItem('maSeanceId');
    if (maSeanceIdToken) {
        const parsedUsertoken: UserToken = JSON.parse(maSeanceIdToken);
        config.headers.Authorization = `Bearer ${parsedUsertoken.access_token}`;
    }
    return config;
});

/**
 * Set global behavior for responses
 */
axiosInstance.interceptors.response.use(
    response => handleResponse(response),
    error => handleError(error)
);

export const handleError = (error: any) => {
    if (error.response.status === 401 || error.response.status === 403) {
        throw new Response(error.response.data, {
            status: error.response.status,
            statusText: error.response.statusText
        });
    }
    toast.error("Une erreur s'est produite.");
    if (error.response) {
        throw new Response(error.response.data, {
            status: error.response.status,
            statusText: error.response.statusText
        });
    }
};

export const handleResponse = (response: AxiosResponse) => {
    return response.data;
};
