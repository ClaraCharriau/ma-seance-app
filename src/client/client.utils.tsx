import { AxiosResponse } from 'axios';

export const HOST = 'http://localhost:7878';

export const handleError = (error: any) => {
    if (error.response) {
        throw new Response(error.response.data, {
            status: error.response.status,
            statusText: error.response.statusText
        });
    }
    console.error(error.message);
    throw error;
};

export const handleResponse = (response: AxiosResponse) => {
    return response.data;
};

export const handleDeleteResponse = (response: AxiosResponse) => {
    return response;
};
