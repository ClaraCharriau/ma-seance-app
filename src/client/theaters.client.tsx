import axios from 'axios';
import { HOST, handleError, handleResponse } from './client.utils';
import { Theater } from '../models/Theater';

const theaters_path = HOST + '/theaters';

export const getTheaterById = async (theaterId: number): Promise<Theater> => {
    return await axios
        .get(theaters_path + '/' + theaterId)
        .then(response => handleResponse(response))
        .catch(error => handleError(error));
};
