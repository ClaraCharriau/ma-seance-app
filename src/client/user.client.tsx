import { Theater } from '../models/Theater';
import axios from 'axios';
import { handleError, handleResponse } from './client.utils';

const host = 'http://localhost:7878';
const users_path = host + '/users';

export const getUserFavTheaters = async (id: number): Promise<Theater[]> => {
    return await axios
        .get(users_path + '/' + id + '/fav-theaters')
        .then(response => handleResponse(response))
        .catch(error => handleError(error, 'GET /fav-theaters'));
};
