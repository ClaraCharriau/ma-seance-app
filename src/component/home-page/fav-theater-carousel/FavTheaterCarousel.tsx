import { useEffect, useState } from 'react';
import { getUserFavTheaters } from '../../../client/user.client';
import { Theater } from '../../../models/Theater';
import { User } from '../../../models/User';
import FavTheaterCard from './FavTheaterCard';
import { CarouselProvider, Slide, Slider } from 'pure-react-carousel';
import style from './FavTheater.module.css';
import { useMediaQuery } from 'react-responsive';

interface FavTheaterCarouselProps {
    currentUser: User;
}

const FavTheaterCarousel = (props: FavTheaterCarouselProps) => {
    const { currentUser } = props;
    const [favTheaters, setFavTheaters] = useState<Theater[]>([]);

    useEffect(() => {
        getUserFavTheaters(currentUser.id).then(theaters => {
            setFavTheaters(theaters);
        });
    }, [currentUser.id]);

    const isDesktop = useMediaQuery({
        query: '(min-device-width: 1024px)'
    });
    const isWideDesktop = useMediaQuery({
        query: '(min-device-width: 1524px)'
    });
    const isTablet = useMediaQuery({ query: '(min-device-width: 300px)' });
    const slideCount = isWideDesktop ? 5 : isDesktop ? 4 : isTablet ? 3 : 2;

    return (
        <CarouselProvider
            visibleSlides={slideCount}
            totalSlides={favTheaters.length}
            step={1}
            naturalSlideWidth={110}
            naturalSlideHeight={100}
            className={style.carousel}
        >
            <Slider className={style.slider}>
                {favTheaters.map(theater => (
                    <Slide index={theater.id} key={theater.id} className={style.slide}>
                        <FavTheaterCard theater={theater} key={theater.id} />
                    </Slide>
                ))}
            </Slider>
        </CarouselProvider>
    );
};

export default FavTheaterCarousel;
