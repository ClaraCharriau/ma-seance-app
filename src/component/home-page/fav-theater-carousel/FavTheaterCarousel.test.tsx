import { render, waitFor } from '@testing-library/react';
import FavTheaterCarousel from './FavTheaterCarousel';
import { getUserFavTheaters } from '../../../client/users/user.client';
import { BrowserRouter } from 'react-router-dom';

jest.mock('../../../client/users/user.client', () => ({
    getUserFavTheaters: jest.fn()
}));
const mockGetUserFavTheaters = getUserFavTheaters as jest.MockedFunction<typeof getUserFavTheaters>;

describe('Fav Theater Carousel Component', () => {
    it('renders Fav Theater Carousel component', async () => {
        const mockUser = {
            id: 1,
            pseudo: 'Jane',
            email: 'test@mail.com'
        };
        const mockFavTheaters = [
            {
                id: 1,
                name: 'Le Grand Rex',
                address: '',
                imgPath: ''
            },
            {
                id: 2,
                name: 'Path√© Levallois',
                address: '',
                imgPath: ''
            },
            {
                id: 3,
                name: 'Super cinoche',
                address: '',
                imgPath: ''
            }
        ];
        mockGetUserFavTheaters.mockResolvedValueOnce(mockFavTheaters);

        const component = render(
            <BrowserRouter>
                <FavTheaterCarousel currentUser={mockUser} />
            </BrowserRouter>
        );

        await waitFor(() => {
            mockFavTheaters.forEach(theater => {
                expect(component.getByText(theater.name)).toBeInTheDocument();
            });
        });
        expect(component.container).toMatchSnapshot();
    });
});
