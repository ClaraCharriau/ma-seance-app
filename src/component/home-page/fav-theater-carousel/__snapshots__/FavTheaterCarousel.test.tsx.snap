// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fav Theater Carousel Component renders Fav Theater Carousel component 1`] = `
<div>
  <div
    class="carousel carousel"
  >
    <div
      aria-label="slider"
      aria-live="polite"
      class="horizontalSlider___281Ls carousel__slider carousel__slider--horizontal slider"
      role="listbox"
      tabindex="0"
    >
      <div
        class="carousel__slider-tray-wrapper carousel__slider-tray-wrap--horizontal"
      >
        <div
          class="sliderTray___-vHFQ sliderAnimation___300FY carousel__slider-tray carousel__slider-tray--horizontal"
          style="width: 150%; transform: translateX(0%) translateX(0px); flex-direction: row;"
        >
          <div
            aria-label="slide"
            aria-selected="true"
            class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--visible slide"
            role="option"
            style="width: 33.333333333333336%; padding-bottom: 30.303030303030305%;"
            tabindex="0"
          >
            <div
              class="slideInner___2mfX9 carousel__inner-slide"
            >
              <a
                class="card"
                href="/theaters/1"
              >
                <img
                  alt="photo de Le Grand Rex"
                  class="theaterCardImg"
                  src="/assets/img/theaters.jpg"
                />
                <div
                  class="theaterName"
                >
                  <p>
                    Le Grand Rex
                  </p>
                </div>
              </a>
            </div>
          </div>
          <div
            aria-label="slide"
            aria-selected="false"
            class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--hidden slide"
            role="option"
            style="width: 33.333333333333336%; padding-bottom: 30.303030303030305%;"
            tabindex="-1"
          >
            <div
              class="slideInner___2mfX9 carousel__inner-slide"
            >
              <a
                class="card"
                href="/theaters/2"
              >
                <img
                  alt="photo de Pathé Levallois"
                  class="theaterCardImg"
                  src="/assets/img/theaters.jpg"
                />
                <div
                  class="theaterName"
                >
                  <p>
                    Pathé Levallois
                  </p>
                </div>
              </a>
            </div>
          </div>
          <div
            aria-label="slide"
            aria-selected="false"
            class="slide___3-Nqo slideHorizontal___1NzNV carousel__slide carousel__slide--hidden slide"
            role="option"
            style="width: 33.333333333333336%; padding-bottom: 30.303030303030305%;"
            tabindex="-1"
          >
            <div
              class="slideInner___2mfX9 carousel__inner-slide"
            >
              <a
                class="card"
                href="/theaters/3"
              >
                <img
                  alt="photo de Super cinoche"
                  class="theaterCardImg"
                  src="/assets/img/theaters.jpg"
                />
                <div
                  class="theaterName"
                >
                  <p>
                    Super cinoche
                  </p>
                </div>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
